src/bluesky/simulators.py

```diff
@@ -104,3 +106,195 @@ async def check_limits_async(plan):
                 ignore.append(obj)
+
+
+class RunEngineSimulator:
+    """This class facilitates testing of a Bluesky plan by recording bluesky messages and
+     injecting responses according to the bluesky Message Protocol (see bluesky docs for details).
+
+    Basic usage consists of
+    1) Registering various handlers to respond to anticipated messages in the experiment plan and fire any
+    needed callbacks.
+    2) Calling simulate_plan()
+    3) Examining the returned message list and making asserts against them
+
+    An example usage:
+    >>> def my_plan():
+    ...     pass
+
+    >>> sim = RunEngineSimulator()
+    >>> messages = sim.simulate_plan(my_plan())
+    >>>
+
+
+    """
+
+    GROUP_ANY = "any"
+
+    def __init__(self):
+        self.message_handlers = []
+        self.callbacks = {}
+        self.next_callback_token = 0
+        self.return_value: Any
+
+    def add_handler_for_callback_subscribes(self):
+        """Add a handler that registers all the callbacks from subscribe messages so we can call them later.
+        You will need to call this as one of the first things if you wish to fire callbacks from the simulator.
+        """
+        self.message_handlers.append(
+            _MessageHandler(
+                lambda msg: msg.command == "subscribe",
+                lambda msg: self._add_callback(msg.args),
+            )
+        )
+
+    def add_handler(
+        self,
+        handler: Callable[[Msg], object],
+        commands: Union[str, Sequence[str]],
+        obj_name: Optional[str] = None,
+        predicate: Optional[Callable[[Msg], bool]] = None,
+    ):
+        """Add the specified handler for a particular message. The handler is prepended so that
+        newer handlers override older ones.
+        Args:
+            handler: a lambda that accepts a Msg and returns an object; the object is sent to the current yield
+            statement in the generator, and is used when reading values from devices, the structure of the
+            object depends on device hinting.
+            commands: the command name for the message as defined in bluesky Message Protocol, or a sequence if
+            more than one matches.
+            obj_name: the name property of the obj to match, can be None (the default) as not all messages have
+            a name.
+            predicate: if specified, then the predicate is used to select the handler rather than the object name.
+        """
+        if isinstance(commands, str):
+            commands = [commands]
+
+        self.message_handlers.insert(
+            0,
+            _MessageHandler(
+                lambda msg: msg.command in commands
+                and (
+                    (predicate and predicate(msg)) or (obj_name is None or (msg.obj and msg.obj.name == obj_name))
+                ),
+                handler,
+            ),
+        )
+
+    def add_read_handler(self, obj_name, value, dict_key="values"):
```

Comment:
I've tried to make the signature of this method such that it results in the least amount of typing in most use cases. One of the things that is kind of annoying at the moment is that the value required to specify for `dict_key` whether it's "values" or the object name is not obvious (to me at any rate) and usually determined by trial and error looking at the message stream. I would much prefer it if the caller never has to specify it and only has to specify the value, which is usually (but not always) just a scalar.

So I would be against making the basic use case require more arguments, and would prefer if there was even some way it could be automatically determined (or can we even safely ignore it and check both?). But we could make it an option to return the entire dictionary if needed.

As far as including the timestamp, I think it's a bad idea to include in unit tests as in most cases the timestamp is not sensitive, and in those cases we do care, in unit tests you almost always want to be able to specify the exact value of the timestamp.

Reply:
I expect that this code will be read more often than it will be written (in the context of "this test is now failing? why is it failing? what is it even doing?!) so I think we should prioritize clarity on reading someone else's tests over brevity of input.

I think returning data structures with different sets of keys is a pretty big divergence from the actual RE.  Some strawman code that would pass without `'timestamps'` but fail in actual usage is

```python
def sneak_plan(obj):
    ret = yield from bps.read(obj)
    for k, v in ret.items:
        # yes, this is a bad idea
        reading, = v.values()
```

Going back to a concern I raised earlier, in the early days we had a shadow-ophyd in bluesky that we used for testing.  The 4th or 5th time we had all our tests passing and then discovered that there was some subtle implementation difference between actual ophyd and our test-ophyd when running on the floor we gave up and just accepted a test-dependency on ophyd.

Reply:
It now returns an instance of Reading with the timestamp initialised to the current time